broker-core/src/main/java/io/zeebe/broker/workflow/state/VariablesState.java

```diff
@@ -0,0 +1,380 @@
+      final Tuple<DirectBuffer, DirectBuffer> currentVariable = localVariableIterator.next();
+
+      if (filter.test(currentVariable.getLeft())) {
+        variableConsumer.accept(currentVariable.getLeft(), currentVariable.getRight());
+
+        if (completionCondition.getAsBoolean()) {
+          return true;
+        }
+      }
+    }
+
+    localVariableIterator.close();
+
+    return false;
+  }
+
+  public void removeAllVariables(long scopeKey) {
+    localVariableIterator.init(scopeKey);
+
+    while (localVariableIterator.hasNext()) {
+      localVariableIterator.next();
+      localVariableIterator.remove();
+    }
+
+    localVariableIterator.close();
+  }
+
+  private void removeVariableLocal(long scopeKey, DirectBuffer name) {
+    int keyOffset = 0;
+    stateKeyBuffer.putLong(keyOffset, scopeKey, ZeebeStateConstants.STATE_BYTE_ORDER);
+    keyOffset += BitUtil.SIZE_OF_LONG;
+    stateKeyBuffer.putBytes(keyOffset, name, 0, name.capacity());
+
+    final int keyLength = keyOffset + name.capacity();
+
+    System.out.println("Removing " + name.getStringWithoutLengthUtf8(0, name.capacity()));
+
+    stateController.remove(variablesHandle, stateKeyBuffer.byteArray(), 0, keyLength);
+  }
+
+  private class LocalVariableIterator implements Iterator<Tuple<DirectBuffer, DirectBuffer>> {
+    private final MutableDirectBuffer scopeKeyBuffer =
+        new UnsafeBuffer(new byte[BitUtil.SIZE_OF_LONG]);
+    private long scopeKey;
+    private RocksIterator rocksDbIterator;
+
+    private final MutableDirectBuffer currentScopeKeyBuffer = new UnsafeBuffer(0, 0);
+
+    private UnsafeBuffer nameView = new UnsafeBuffer(0, 0);
+    private UnsafeBuffer valueView = new UnsafeBuffer(0, 0);
+    private Tuple<DirectBuffer, DirectBuffer> value = new Tuple<>(nameView, valueView);
+
+    private void init(long scopeKey) {
+      this.scopeKey = scopeKey;
+      scopeKeyBuffer.putLong(0, scopeKey, ZeebeStateConstants.STATE_BYTE_ORDER);
+
+      rocksDbIterator = stateController.getDb().newIterator(variablesHandle);
+      rocksDbIterator.seek(scopeKeyBuffer.byteArray());
+    }
+
+    @Override
+    public boolean hasNext() {
+      if (rocksDbIterator.isValid()) {
+        final byte[] currentKey = rocksDbIterator.key();
+        currentScopeKeyBuffer.wrap(currentKey);
+
+        return currentScopeKeyBuffer.getLong(0, ZeebeStateConstants.STATE_BYTE_ORDER) == scopeKey;
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public Tuple<DirectBuffer, DirectBuffer> next() {
+
+      final byte[] currentKey = rocksDbIterator.key();
+      final byte[] currentValue = rocksDbIterator.value();
+
+      // Must make copies because #next invalidates the current key and value;
+      // This could probably be avoided if RocksIteratorInterface exposed a #hasNext method, in
+      // which case
+      // we wouldn't have to move to the next element already here
+      final int nameLength = currentKey.length - VARIABLES_NAME_OFFSET;
```

Comment:
I'm not 100% happy with this solution, because RocksDB also copies already and allocates these byte arrays, then we copy again and on `getVariablesAsDocument` again. Maybe we find a better solution?